Show running workflow error logs

SuaYoo opened the pull request:
Resolves https://github.com/webrecorder/browsertrix-cloud/issues/1187

<!-- Fixes #issue_number -->

### Changes

- Adds "Logs" tab to workflow detail
- Shows error logs in expandable section in "Watch" tab
- Show corresponding message (no logs yet or logs temporarily unavailable) when `/errors` returns 503

### Screenshots

| Page | Image/video |
| ---- | ----------- |
| Workflow Detail - Logs (never crawled) | <img width="920" alt="Screenshot 2023-09-27 at 2 27 03 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/5d201b5c-1546-4751-bd9f-24637d9ec371"> |
| Workflow Detail - Logs (no logs) | <img width="920" alt="Screenshot 2023-10-02 at 3 44 23 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/004be3fd-b0d6-4322-aa7b-0c94a07942b8"> |
| Workflow Detail - Logs (with logs) | <img width="930" alt="Screenshot 2023-09-27 at 2 32 55 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/a328e919-bd90-445d-987d-0aa1a3aadb4c"> |
| Workflow Detail - Logs (currently running) | <img width="924" alt="Screenshot 2023-09-27 at 2 41 14 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/15f993a8-2d40-4117-ae74-6265ad9174ae"> |
| Workflow Detail - Watch | <img width="932" alt="Screenshot 2023-09-27 at 2 51 31 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/170712a8-e62c-47a5-8c89-8470f9664883"> | 


<!-- ### Follow-ups -->

SuaYoo committed:
add crawl logs tab

SuaYoo committed:
show logs conditionally

SuaYoo committed:
show log in watch view

tw4l commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -119,6 +124,7 @@ export class WorkflowDetail extends LiteElement {
   private readonly tabLabels: Record<Tab, string> = {
     crawls: msg("Crawls"),
     watch: msg("Watch Crawl"),
+    logs: msg("Logs"),
```
This is "Error Logs" in the crawl - might be good to be consistent for now until we display full logs?

tw4l commented on frontend/src/pages/org/crawl-detail.ts:
```diff
@@ -893,7 +894,7 @@ ${this.crawl?.description}
-      this.logs = await this.getCrawlLogs(params);
+      this.logs = await this.getCrawlErrors(params);
     } catch {
       this.notify({
```
It might be good to only display this if it's not a 503? Otherwise the notification right when the crawl kicks off might be a bit concerning, even though it's expected.

tw4l reviewed (commented):
A few small notes left in-line. I'm also noticing that the expanded Error Logs card closes periodically, maybe when polling happens?

tw4l commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -497,6 +503,19 @@ export class WorkflowDetail extends LiteElement {
+          ?disabled=${!this.logs?.total}
+        >
+          <sl-icon slot="prefix" name="download"></sl-icon>
+          ${msg("Download Logs")}
```
We might want to be clearer that this is the logs of the last finished crawl?

tw4l replied:
Doesn't have to be on the button necessarily, could display the crawl id or a link to the crawl at the top, something like that?

SuaYoo replied:
Is the `/logs` only available for finished crawls? I assumed you could download logs for the currently running logs. If not, we could also disable this button so it's not confusing as to which crawl logs you're downloading, since you can always go to the last finished crawl's error logs tab to download them.

SuaYoo replied:
Went ahead with disabling the button in the meantime:
<img width="330" alt="Screenshot 2023-10-02 at 3 42 44 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/4672952/7687a94f-cb38-4a20-9213-a32591ccd427">

tw4l commented:
I'm also seeing some funny behaviors with the error logs display:
- It seems like new errors only display after a page refresh
- When I start a new crawl from the same workflow, the errors from the last crawl are still visible until I refresh the page

Maybe this is expected? I assumed coming in that there would be some polling to ensure the new errors are immediately visible without a page reload.

SuaYoo committed:
update labels

SuaYoo committed:
handle 503

SuaYoo committed:
fetch error logs when crawl is running

SuaYoo committed:
disable download button

SuaYoo committed:
update no logs message

SuaYoo commented:
> I'm also seeing some funny behaviors with the error logs display

Should be fixed, I wasn't accounting for "run now" properly.

> I'm also noticing that the expanded Error Logs card closes periodically, maybe when polling happens?

I haven't been able to reproduce this, is there a workflow in your org that I can repro with?

ikreymer commented:
> > I'm also noticing that the expanded Error Logs card closes periodically, maybe when polling happens?
> 
> I haven't been able to reproduce this, is there a workflow in your org that I can repro with?

yeah, also haven't been able to repro this so far. Looks great overall, will make debugging so much easier!
I think this should be good to go!

ikreymer commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -1053,10 +1045,114 @@ export class WorkflowDetail extends LiteElement {
+                    class="border rounded-lg p-4 flex flex-col items-center justify-center"
+                  >
+                    <p class="text-center text-neutral-400">
+                      ${msg("No logs found for latest crawl.")}
```
```suggestion
                      ${msg("No errors found for latest crawl.")}
```

ikreymer commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -1053,10 +1045,114 @@ export class WorkflowDetail extends LiteElement {
+          () => html`<div class="mb-4">
+            <btrix-alert variant="success" class="text-sm">
+              ${msg(
+                html`Viewing logs for currently running crawl.
```
```suggestion
                html`Viewing error logs for currently running crawl.
```

ikreymer committed:
text / error message fixes:
- add 'error logs' instead of 'logs' to be more clear
- if no logs and status is 'waiting_capacity', indicate error logs not available temporarily, otherwise indicate no error logs yet
- for waiting_capacity status, change 'crawl start' -> 'crawl continue' in log message
- backend: return 'error_logs_not_available' instead of 'redis_connection_error' along with 503 for unavailable logs (fixes #1238)

ikreymer commented:
Also added some text tweaks, changed backend 503 return value to fix #1238
